Feat/save task locally

justdvnsh opened the pull request:
### Description
Users can look at their task with this PR, even when they are offline. The tasks and relationship details are stored in the backend. Current relation details needed to be stored as well, since the Relation and Task Fragment rely on the current relation as an input parameter. One thing to note. __In the database, I have allowed the queries to run on the main thread. However, this should not be done in production. I did it just for testing purpose, since, not running it on main thread would cost us one more dependency + a lot more code. But since we are just storing the current relation and tasks, we can easily get away with running it on main thread__

Fixes #937 

### Type of Change:

- Code
- Quality Assurance

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
I have attached the working screen recorded video of the issue.
[https://drive.google.com/file/d/1YPc6rVQjeirug6Bma2E0W6mlXUHakegc/view?usp=sharing](https://drive.google.com/file/d/1YPc6rVQjeirug6Bma2E0W6mlXUHakegc/view?usp=sharing)

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

justdvnsh committed:
feat: Added dependencies for room - Issue #937

justdvnsh committed:
feat: Added Room Database, entities and Daos

justdvnsh committed:
feat: Added Database and DAOs for database related quesries

justdvnsh committed:
feat: Added Feature to save tasks locally. Fix Issue - #937

body: Added feature through which user can view the tasks locally, without internet. However, they cannot add or update the task.

justdvnsh committed:
refactor: Minor Fix

CodeChamp-SS commented:
@justdvnsh how will it handle the case when the user signs out and logins with another account ? i think the tasks from the previous account will also be shown as they are not getting deleted from the db :-)

justdvnsh commented:
@CodeChamp-SS Yes, you are right .
One thing that can be done is that I can add an sql query to fetch the task only for a particular user .
Thanks for pointing this out. It completely skipped my mind !

CodeChamp-SS commented:
> @CodeChamp-SS Yes, you are right .
> One thing that can be done is that I can add an sql query to fetch the task only for a particular user .
> Thanks for pointing this out. It completely skipped my mind !

np 👍
btw you can also try deleting the tasks and relations of a user when he/she signs out

justdvnsh committed:
refactor: Refactored code to query tasks for only the current user.

body: Added relationId parameter in the TASK model, so that the queries can be made to fetch the task only for one particular user.

justdvnsh commented:
@CodeChamp-SS Yeah, but then the next time he/she logs in, they wouldn't be having the relations and tasks saved in offline mode. I have made changes so that , all the relations and tasks for all the accounts of a user can be stored and then retrieved for current account in usage.

iofall reviewed (dismissed):
Looks good to me! Tested locally as well 🎉

rpattath commented:
@iofall do you have any screenshots from when you tested locally? If so, we could skip `Needs Testing` for this PR.

iofall commented:
> @iofall do you have any screenshots from when you tested locally? If so, we could skip `Needs Testing` for this PR.

I don't have them, I will test it again and send the screenshots, as soon as possible

iofall commented:
Screenshots of Local Testing:

The normal, current version of the app does not save the tasks locally, so if the user does not have network access, they cannot look at their tasks.

**Current Version**
When no internet connection is present, the relation screen is not rendered
<img src="https://user-images.githubusercontent.com/50991099/97794885-968b2080-1c25-11eb-8b00-f0a5b02426fd.png" width="25%">

**This PR**
When no internet connection is present, the relation screen is still rendered from the data that was saved locally

<img src="https://user-images.githubusercontent.com/50991099/97795014-1796e780-1c27-11eb-85d4-af03d287b033.png" width="25%"> <img src="https://user-images.githubusercontent.com/50991099/97794704-575bd000-1c23-11eb-8451-91c894463cf4.png" width="25%">

aditmehta9 commented:
@justdvnsh can you please solve the merge conflicts?

epicadk commented:
@justdvnsh please solve the merge conflicts

justdvnsh committed:
Merge branch 'develop' into feat/save_task_locally

justdvnsh committed:
updated task.kt 

resolved merge conflicts

justdvnsh committed:
updated relation pager fragment

justdvnsh committed:
updated task fragment

justdvnsh commented:
@epicadk DONE

vj-codes reviewed (changes requested):
@justdvnsh please resolve the merge conflicts

vj-codes reviewed (changes requested):
@justdvnsh closing this due to inactivity, please let us know if you would like to continue again.
Thank you for your contribution :)
